src/stream/src/executor/backfill/no_shuffle_backfill.rs

```diff
@@ -485,29 +502,43 @@ where
                         )
                         .await?;
                         tracing::trace!(
-                            epoch = ?barrier.epoch,
+                            ?epoch,
                             ?current_pos,
                             total_snapshot_processed_rows,
                             "Backfill position persisted after completion"
                         );
                     }
-
-                    // For both backfill finished before recovery,
-                    // and backfill which just finished, we need to update mview tracker,
-                    // it does not persist this information.
-                    self.progress
-                        .finish(barrier.epoch.curr, total_snapshot_processed_rows);
-                    tracing::trace!(
-                        epoch = ?barrier.epoch,
-                        "Updated CreateMaterializedTracker"
-                    );
                     yield msg;
                     break;
                 }
                 yield msg;
             }
         }
 
+        // NOTE(kwannoel):
+        // Progress can only be finished after at least 1 barrier.
+        // This is to make sure that downstream state table is flushed,
+        // before the mview is made visible.
+        // Otherwise the mview could be inconsistent with upstream.
+        // It also cannot be immediately `finished()` after yielding the barrier,
+        // by using the current_epoch of that barrier, since that will now be the previous epoch.
+        // When notifying the global barrier manager, local barrier manager always uses the
+        // current epoch, so it won't see the `finished` state when that happens,
+        // leading to the stream job never finishing.
+        // Instead we must wait for the next barrier, and finish the progress there.
```

Comment:
Full explanation of the bug.

Reply:
How did you validate this bug occurred before your fix? IIRC, the `progress.finish()` only updates the backfill progress in a local barrier manager and the local barrier manager sends the progress to meta only when the barrier is collected so it will be fine.

Reply:
+1

Reply:
Interesting. It is caught by my test, reproducible with:
```sh
runtest() { MADSIM_TEST_SEED=1698936328083879000 RUST_LOG="risingwave_meta=debug,risingwave_stream::executor::flow_control=trace,risingwave_stream::executor::backfill::no_shuffle_backfill=trace,integration_tests=debug,risingwave_stream::executor::mview::materialize=trace" ./risedev sit-test test_background_mv_barrier_recovery; }
runtest 1>out.log 2>&1
```

Reply:
I will do further investigation.

Reply:
The observation is that the test `test_background_mv_barrier_recovery` fails this assertion:

```rs

    let t_count = session.run("SELECT COUNT(v1) FROM t").await?;
    let mv1_count = session.run("SELECT COUNT(v1) FROM mv1").await?;

    assert_eq!(t_count, mv1_count);
```

Reply:
LHS is `1000`, RHS is `999`, and the row missing had offset `473`. It was the last row before snapshot read completed.

Reply:
```
2022-06-23T08:46:30.950911Z TRACE risingwave_stream::executor::backfill::no_shuffle_backfill: Backfill has finished, waiting for barrier
2022-06-23T08:46:31.032269Z TRACE risingwave_stream::common::table::state_table: commit state table table_id=1001 epoch=2538785342619648
2022-06-23T08:46:31.032269Z TRACE risingwave_stream::common::table::state_table: commit state table table_id=1002 epoch=2538785342619648
2022-06-23T08:46:31.034755Z TRACE risingwave_stream::common::table::state_table: commit state table table_id=1001 epoch=2538785342619648
2022-06-23T08:46:31.034755Z TRACE risingwave_stream::common::table::state_table: commit state table table_id=1002 epoch=2538785342619648
2022-06-23T08:46:31.034949Z TRACE risingwave_stream::common::table::state_table: commit state table table_id=1001 epoch=2538785342619648
2022-06-23T08:46:31.034950Z TRACE risingwave_stream::common::table::state_table: commit state table table_id=1003 epoch=2538785342619648

2022-06-23T08:46:31.034986Z TRACE risingwave_stream::common::table::state_table: commit state table table_id=1001 epoch=2538785342619648
2022-06-23T08:46:31.034986Z TRACE risingwave_stream::common::table::state_table: commit state table table_id=1002 epoch=2538785342619648
2022-06-23T08:46:31.035345Z TRACE risingwave_stream::common::table::state_table: commit state table table_id=1001 epoch=2538785342619648
2022-06-23T08:46:31.035345Z TRACE risingwave_stream::common::table::state_table: commit state table table_id=1002 epoch=2538785342619648
2022-06-23T08:46:31.036654Z TRACE risingwave_stream::common::table::state_table: commit state table table_id=1001 epoch=2538785342619648
2022-06-23T08:46:31.036654Z TRACE risingwave_stream::common::table::state_table: commit state table table_id=1002 epoch=2538785342619648
2022-06-23T08:46:31.046158Z TRACE risingwave_stream::executor::backfill::no_shuffle_backfill: Backfill position persisted after completion epoch=EpochPair { curr: 2538785408155648, prev: 2538785342619648 } current_pos=Some(OwnedRow([Some(Serial(Serial(162481193722486794)))])) total_snapshot_processed_rows=500
2022-06-23T08:46:31.046158Z TRACE risingwave_stream::executor::backfill::no_shuffle_backfill: Updated CreateMaterializedTracker epoch=EpochPair { curr: 2538785408155648, prev: 2538785342619648 }
2022-06-23T08:46:31.046158Z TRACE risingwave_stream::common::table::state_table: commit state table table_id=1002 epoch=2538785342619648
2022-06-23T08:46:31.046158Z TRACE risingwave_stream::executor::backfill::no_shuffle_backfill: Backfill has already finished and forward messages directly to the downstream
2022-06-23T08:46:31.055281Z DEBUG risingwave_meta::barrier::progress: all actors done for creating mview with table_id 1002!
2022-06-23T08:46:31.071360Z DEBUG risingwave_meta::barrier::recovery: finished stream job 1002
2022-06-23T08:46:31.075987Z DEBUG risingwave_meta::manager::catalog: notifying frontend id=1002
2022-06-23T08:46:31.081673Z DEBUG risingwave_meta::barrier::recovery: notified frontend for stream job 1002
2022-06-23T08:46:31.838873Z  INFO pgwire_query_log: mode=(simple query) session=0 status=ok time=222012ms sql=WAIT;
2022-06-23T08:46:31.888226Z  INFO pgwire_query_log: mode=(simple query) session=0 status=ok time=34ms sql=SELECT COUNT(v1) FROM t
2022-06-23T08:46:31.941147Z  INFO pgwire_query_log: mode=(simple query) session=0 status=ok time=37ms sql=SELECT COUNT(v1) FROM mv1
2022-06-23T08:46:32.027126Z  INFO pgwire_query_log: mode=(simple query) session=0 status=ok time=68ms sql=SELECT v1 FROM t WHERE v1 NOT IN (SELECT v1 FROM mv1)
2022-06-23T08:46:32.030333Z DEBUG integration_tests::recovery::background_ddl: missing_rows="473"
2022-06-23T08:46:32.030915Z TRACE risingwave_stream::common::table::state_table: commit state table table_id=1001 epoch=2538785408155648
2022-06-23T08:46:32.030915Z TRACE risingwave_stream::common::table::state_table: commit state table table_id=1002 epoch=2538785408155648
2022-06-23T08:46:32.031159Z TRACE risingwave_stream::common::table::state_table: commit state table table_id=1001 epoch=2538785408155648
2022-06-23T08:46:32.031159Z TRACE risingwave_stream::common::table::state_table: commit state table table_id=1002 epoch=2538785408155648
2022-06-23T08:46:32.032395Z TRACE risingwave_stream::common::table::state_table: commit state table table_id=1001 epoch=2538785408155648
2022-06-23T08:46:32.032396Z TRACE risingwave_stream::common::table::state_table: commit state table table_id=1002 epoch=2538785408155648
2022-06-23T08:46:32.032507Z TRACE risingwave_stream::common::table::state_table: commit state table table_id=1001 epoch=2538785408155648
2022-06-23T08:46:32.032507Z TRACE risingwave_stream::common::table::state_table: commit state table table_id=1002 epoch=2538785408155648
2022-06-23T08:46:32.035839Z TRACE risingwave_stream::common::table::state_table: commit state table table_id=1001 epoch=2538785408155648
2022-06-23T08:46:32.035839Z TRACE risingwave_stream::common::table::state_table: commit state table table_id=1002 epoch=2538785408155648
2022-06-23T08:46:32.036764Z TRACE risingwave_stream::common::table::state_table: commit state table table_id=1001 epoch=2538785408155648
2022-06-23T08:46:32.036765Z TRACE risingwave_stream::common::table::state_table: commit state table table_id=1002 epoch=2538785408155648
2022-06-23T08:46:32.057247Z DEBUG risingwave_meta::hummock::metrics_utils: LSM Compacting STAT {}
2022-06-23T08:46:32.057247Z DEBUG risingwave_meta::hummock::metrics_utils: LSM Compacting STAT {}
2022-06-23T08:46:32.109910Z  INFO pgwire_query_log: mode=(simple query) session=0 status=ok time=71ms sql=SELECT _row_id, v1 FROM t WHERE v1 NOT IN (SELECT v1 FROM mv1)
2022-06-23T08:46:32.113030Z DEBUG integration_tests::recovery::background_ddl: missing_rows_with_row_id="162481193722486794 473"
```

Reply:
<img width="1493" alt="Screenshot 2023-11-03 at 12 54 29 PM" src="https://github.com/risingwavelabs/risingwave/assets/47273164/a89d6dc5-3ada-4188-bbb6-e8216600c659">

Seems like the state table only committed notified frontend.

Reply:
Can ignore, it's a false positive, because the epoch is from `prev`, so actually the ordering is correct. The commits are from the next barrier, rather than the current one.

Reply:
The root cause is because recovered stream job requires checkpoint barrier: https://github.com/risingwavelabs/risingwave/pull/13219/commits/058b343d63f1571f7dd8c95515852fa500dd8aac

Reply:
> The root cause is because recovered stream job requires checkpoint barrier: [058b343](https://github.com/risingwavelabs/risingwave/commit/058b343d63f1571f7dd8c95515852fa500dd8aac)

This is a pretty good catch!

Reply:
Without checkpoint barrier, once the stream job becomes exposed to frontend, the batch query will read from the latest checkpoint barrier.

Now upstream data could have been checkpointed several epoch's ago, so it will return the full data.
But the latest snapshot data being streamed into the mv might not be checkpointed yet. So it will not be visible, leading to the inconsistency between upstream and downstream mv.